src/main/java/org/eclipse/yasson/internal/serializer/ContainerDeserializer.java

```diff
@@ -0,0 +1,322 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v1.0 and Eclipse Distribution License v. 1.0
+ * which accompanies this distribution.
+ * The Eclipse Public License is available at http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * Contributors:
+ * Tomas Kraus
+ ******************************************************************************/
+package org.eclipse.yasson.internal.serializer;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import javax.json.bind.JsonbException;
+import javax.json.bind.serializer.DeserializationContext;
+import javax.json.bind.serializer.JsonbDeserializer;
+import javax.json.stream.JsonParser;
+
+import org.eclipse.yasson.internal.JsonbContext;
+import org.eclipse.yasson.internal.JsonbParser;
+import org.eclipse.yasson.internal.JsonbRiParser;
+import org.eclipse.yasson.internal.ReflectionUtils;
+import org.eclipse.yasson.internal.RuntimeTypeInfo;
+import org.eclipse.yasson.internal.Unmarshaller;
+import org.eclipse.yasson.internal.model.ClassModel;
+import org.eclipse.yasson.internal.properties.MessageKeys;
+import org.eclipse.yasson.internal.properties.Messages;
+
+/**
+ * Internal container de-serializing interface.
+ *
+ * @param <T> container type
+ */
+interface ContainerDeserializer<T> extends JsonbDeserializer<T> {
```

Comment:
It is needed, but not at such a wide scope.  Moved to `MapSerializer.Delegate<K,V>`.

Reply:
Anyway, we had a long and challenging discussion on this topic and I agree that it can be removed now and introduced later when it will be used more than once.
Because of this, I would like to merge this PR without squashing.